### ISSUE TITLE
Get parent/children options for VMs

### ISSUE THREAD

**Issue by rvsia** [ManageIQ/manageiq-api#637]
**Description**

We are converting `Virtual Machine edit form` to data driven forms and it would be nice to have an API endpoint, which returns all VMs which can be selected as a child or parent for the edited VM.

**Links to code in UI**
https://github.com/ManageIQ/manageiq-ui-classic/blob/d1822704512634446b858017adf999e560d7c850/app/controllers/vm_common.rb#L1353 https://github.com/ManageIQ/manageiq-ui-classic/blob/a459a4823a30b138f44a412b8c10a1f969b1b19a/app/controllers/vm_common.rb#L898

Cc @gtanzillo @hchrast

**Comment by lpichler** [comment]
@miq-bot assign @lpichler

**Comment by lpichler** [comment]
after discussion with @rvsia we need to add new endpoint for this. there is no possibility to use API filtering (no nested condition and GET request query limit)

@miq-bot unassign @lpichler 

I heard that @d-m-u is going to working on It ? 😊 if not, feel free to assign it to me 🏇

**Comment by d-m-u** [comment]
@miq-bot assign @d-m-u

**Comment by d-m-u** [comment]
So I would be shocked if there weren't an existing way to do this that doesn't involve the overhead of _adding a new endpoint_ for something that seems trivial.

**Comment by lpichler** [comment]
@d-m-u I don't want to shock you  :D but I am afraid that we need new endpoint because are API doesn't support nested expressions 

we need to get vms  which you gen get by method [parent_choices](https://github.com/ManageIQ/manageiq-ui-classic/blob/a459a4823a30b138f44a412b8c10a1f969b1b19a/app/controllers/vm_common.rb#L898)

API request could like:

`GET /api/vms/:id/parent_choices` (maybe better name)

and you can see that in the in method is more queries
```
...
 ems_ids = VmOrTemplate.where(:id => ids).distinct.pluck(:ems_id).compact
      parent_item_scope = Rbac.filtered(VmOrTemplate.where(:ems_id => ems_ids).where.not(:id => ids).order(:name))
...
```

so It seem to me that we are not able to do it with existing possibilities in API.
Anyway if you know how let us know,

thanks!

**Comment by Fryguy** [comment]
Can you explain the feature here? What is a parent/child VM?  Do you mean genealogy? If so, do we support this in the current UI, because this feels like a feature that should not exist as it could be easy for a user to create a bad connection. cc @agrare

**Comment by rvsia** [comment]
@Fryguy If you go in the UI to Virtual Machines and select one and in the configuration click on Edit VM, you can see a form, where you set: a custom identifier, description, parent VM and child VMs. Actually it’s on more places in the UI - templates, volume_templates (instances, workloads in UI.) API endpoint which describes what you can change: http://manageiq.org/docs/reference/fine/api/reference/vms#edit-vm
(We would need probably to add also all other types of entities to the API (get, parent choices and edit endpoints), which share this form: templates, volume_templates and maybe some more.)

**Comment by rvsia** [comment]
cc @martinpovolny

**Comment by martinpovolny** [comment]
@Fryguy, @agrare : any progress on this?

**Comment by agrare** [comment]
@martinpovolny any response to

> because this feels like a feature that should not exist as it could be easy for a user to create a bad connection
?


This doesn't seem like something that we even need

**Comment by martinpovolny** [comment]
What is wrong with the answer https://github.com/ManageIQ/manageiq-api/issues/637#issuecomment-518808941 where @rvsia points to a place in the UI where we are actually using this?

Does that mean that the feature can be removed fromt he UI?

**Comment by agrare** [comment]
> What is wrong with the answer #637 (comment) where @rvsia points to a place in the UI where we are actually using this?

That doesn't answer if we need it, just that we happen to currently do it

> Does that mean that the feature can be removed from the UI?

That's what we're wondering.  I can't think of a use-case for why a user would need to manually modify this.  We set it automatically when provisioning from a template.  Outside of that I'm not sure why you would want to set it yourself.

**Comment by martinpovolny** [comment]
Ok, understood. Neither can I, nor can @rvsia or @lpichler.

Who can tell that? Who is able to make a decision about removing the option from the UI?

**Comment by agrare** [comment]
@Loicavenel can you help here?

**Comment by Loicavenel** [comment]
@agrare always asked myself what is this feature about.. I will suspect this is after discovery you can create Genealogy.. I did test it and this is exactly what is happening... Difficult to remove the feature as he may be used by customers

**Comment by martinpovolny** [comment]
Thank you @Loicavenel!

@agrare : is that enough to justify this PR?

(My motivation for pushing this is progressing with the rewrite of the UI as a SPA in React and Patternfly and UI work is blocked on this.)

**Comment by agrare** [comment]
Okay, if we can't remove it then getting back to the original question...if we're letting users manually set parent/child I don't know that we can scope by anything.  The possibilities might just be all vms_and_templates on the ems.

**Comment by martinpovolny** [comment]
@lpichler, @rvsia : can you answer that? To me it seems that yes, the possibility is everything on the set of parent emss except for the vms that are being edited.

I read @lpichler 's response: https://github.com/ManageIQ/manageiq-api/issues/637#issuecomment-516828298 as it is a performance issues? Multiple API queries would be needed?

@d-m-u indicated it should be simple https://github.com/ManageIQ/manageiq-api/issues/637#issuecomment-517451576, but did not provide more details.

We need to talk more to move forward with this ;-)

**Comment by rvsia** [comment]
@martinpovolny I am not sure that I can answer that. I posted the link to the code in UI which provides the options, that’s all I know about it. To be honest, I am little lost in all this - how the entities are different from each other (when they share the exact same code and the form.)

**Comment by lpichler** [comment]
@martinpovolny we were walking thru with @rvsia and we need just endpoint  for parent_choices, as work started here https://github.com/ManageIQ/manageiq-api/pull/671

|Model|GET| Parent Choices| EDIT|Path in UI|
|-|-|-|-|-|
|Vm| 🆗   | ⛔️ | 🆗 |Compute->Clouds->Instances, Compute->Infra->Virtual Machines|
|Template(it is also VolumeTemplate and other descendants)| 🆗   | ⛔️ | 🆗 |Compute->Clouds->Templates, Compute->Infra->Virtual Machines|

**Comment by lpichler** [comment]
in the end we are able to do it thru existing api:
so we method 
```ruby
  def parent_choices(ids)
    @parent_choices = {}
    @parent_choices[Digest::MD5.hexdigest(ids.inspect)] ||= begin
      ems_ids = VmOrTemplate.where(:id => ids).distinct.pluck(:ems_id).compact
      parent_item_scope = Rbac.filtered(VmOrTemplate.where(:ems_id => ems_ids).where.not(:id => ids).order(:name))

      parent_item_scope.pluck(:name, :location, :id).each_with_object({}) do |vm, memo|
        memo[vm[0] + " -- #{vm[1]}"] = vm[2]
      end
    end
  end
```
**1) Parent VM** (in VM edit screen of the *vm* )
```ruby
parent_choices(**[vm.id]**)
```
**2) Available VMS** (in VM edit screen)
```ruby
parent_choices(**[vm.id] + [vm.children]**)
```

so we can convert query from the method:
```ruby
  ems_ids = VmOrTemplate.where(:id => ids).distinct.pluck(:ems_id).compact
  parent_item_scope = Rbac.filtered(VmOrTemplate.where(:ems_id => ems_ids).where.not(:id => ids).order(:name))
```
ems_id is always same because there are only vms from the same provider in the game

```ruby

  Rbac.filtered( ExtManagementSystem.find_by(:id => vm.ems_id).vms.where.not(:id => ids) )
```
so for **1) Parent VM** (in VM edit screen of the *vm* )
ids is [vm.id] and it could be done thru API:

`/api/providers/{vm.ems_id}/vms?expand=resources&attributes=id&filter[]=id!={vm.id}`

and for 
**2) Available VMS** (in VM edit screen)
we need two requests:
**any_vms** = `/api/providers/{vm.ems_id}/vms?expand=resources&attributes=id&filter[]=id!={vm.id}`

**children_vms** = `http://localhost:3090/api/vms/1?attributes=child_resources`

and then we need to remove in javascript **children_vms** from **any_vms**

**available_vms** = **any_vms** - **children_vms** 

@rvsia 

let me know how this works.

NOTE: for templates: /api/providers/1/templates

**Comment by martinpovolny** [comment]
> available_vms = any_vms - children_vms

I wonder if we are dealing with number of VMs that are ok to process at once. Or if some form of pagination or filtering will be needed.

@lpichler : can you make an estimate on number of items we are dealing with here?

**Comment by lpichler** [comment]
> @lpichler : can you make an estimate on number of items we are dealing with here?

we are dealing basically with all VMs of provider(1) without edited VM or (2) VM's children) - so hundreds, thousands...

so pagination and filtering is useful to have.

not sure how what are your plans but I think that select with search would be enough.

**Comment by gtanzillo** [comment]
Closing, as workaround was given with existing API
